### ISSUE TITLE
Incorrect syntax near ';'.

### ISSUE THREAD

**Issue by boginw** [andreisabau/EFBulkInsert#2]
Running Entity Framework 6.2.0 here, and getting syntax error when using the extension method.
Here's the exception:
```
System.Data.SqlClient.SqlException occurred
  HResult=0x80131904
  Message=Incorrect syntax near ';'.
  Source=.Net SqlClient Data Provider
  StackTrace:
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.FillInternal(DataSet dataset, DataTable[] datatables, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataSet dataSet, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataSet dataSet)
   at EFBulkInsert.BulkInsertExtension.MergeDataIntoOriginalTable(DbContext dbContext, EntityMetadata entityMetadata, Int32 startIndex)
   at EFBulkInsert.BulkInsertExtension.BulkInsert[T](DbContext dbContext, IEnumerable`1 entites, Int32 batchSize)
   at Project.Data.Core.Repositories.Repository`1.BulkInsert(IEnumerable`1 toAdd, Int32 batchSize) in C:\ProjectFolder\Data\Core\Repositories\Repository.cs:line 84
   at FakeModels.DBSeeder.Seed() in C:\ProjectFolder\FakeModels\DBSeeder.cs:line 115
   at Project.Program.Main(String[] args) in C:\ProjectFolder\Project\Program.cs:line 23
```

Here's the last run Query, found on SQL Server Profiler 17:
```
MERGE INTO CatalogueLineItem AS DestinationTable
                                                      USING (SELECT * FROM ##TEMP_3e4c2877_9705_4b67_8dfa_5a77bf67d6e1 WHERE ArrayIndex >= 0) AS TempTable
                                                      ON 1 = 2
                                                      WHEN NOT MATCHED THEN INSERT ([CatalogueId],[LineId],[ActionCode],[LifecycleStatusCode],[Note],[Orderable],[OrderableUnit],[ContentUnitQuantity],[OrderQuantityIncrementNumeric],[MinimumOrderQuantity],[MaximumOrderQuantity],[SellerSupplierCustomerAssignedId],[SellerSupplierAdditionalId],[SellerSupplierPartyIdType],[SellerSupplierPartyId],[SellerSupplierPartyName],[LineValidityStartDate],[LineValidityStartTime],[LineValidityEndDate],[LineValidityEndTime],[LeadTimeUnit],[LeadTimeMeasure],[PriceMinimumQuantity],[PriceMaximumQuantity],[PriceAmountCurrency],[PriceAmount],[PriceBaseQuantityUnit],[PriceBaseQuantity],[PriceOrderableUnitFactorRate],[AllowanceChargeIndicator],[AllowanceChargeReasonCode],[AllowanceChargeReason],[AllowanceChargeMultiplierFactorNumeric],[AllowanceChargePrepaidIndicator],[AllowanceChargeAmount],[AllowanceChargeBaseAmount],[DeliveryUnitBatchQuantityUnit],[DeliveryUnitBatchQuantity],[DeliveryUnitConsumerQuantityUnit],[DeliveryUnitConsumerQuantity],[TaxCategoryId],[TaxCategoryName],[TaxCategoryPercent],[TaxCategoryBaseUnit],[TaxPerUnitAmount],[TaxExemptionReasonCode],[TaxExemptionReason],[TaxSchemeId],[TaxSchemeName],[TaxSchemeTypeListId],[TaxSchemeListAgencyId],[TaxSchemeListAgencyName],[TaxScheme],[TaxCurrencyCode],[ItemDescription],[PackQuantity],[PackSize],[ItemName],[HazardousRiskIndicator],[AdditionalInformation],[Keywords],[BrandName],[ModelName],[BuyersItemId],[SellersItemId],[ManufacturersItemId],[StandardItemId],[CatalogueItemId],[AdditionalItemId],[CommodityClassificationListName],[CommodityClassificationListVersion],[CommodityClassificationCode],[ManufacturerPartyIdType],[ManufacturerPartyId],[ManufacturerPartyName],[OriginAddressCountryId],[OriginAddressCountryName],[IsOsdItem],[PackLevel1Unit],[PackLevel1Quantity],[PackLevel1Ean],[PackLevel1Width],[PackLevel1Height],[PackLevel1Length],[PackLevel1GrossWeight],[PackLevel2Unit],[PackLevel2Quantity],[PackLevel2Ean],[PackLevel2Width],[PackLevel2Height],[PackLevel2Length],[PackLevel2GrossWeight],[PackLevel3Unit],[PackLevel3Quantity],[PackLevel3Ean],[PackLevel3Width],[PackLevel3Height],[PackLevel3Length],[PackLevel3GrossWeight],[PackLevel4Unit],[PackLevel4Quantity],[PackLevel4Ean],[PackLevel4Width],[PackLevel4Height],[PackLevel4Length],[PackLevel4GrossWeight],[PackLevelPalletUnit],[PackLevelPalletQuantity],[PackLevelPalletEan],[PackLevelPalletWidth],[PackLevelPalletHeight],[PackLevelPalletLength],[PackLevelPalletGrossWeight]) VALUES 
([CatalogueId],[LineId],[ActionCode],[LifecycleStatusCode],[Note],[Orderable],[OrderableUnit],[ContentUnitQuantity],[OrderQuantityIncrementNumeric],[MinimumOrderQuantity],[MaximumOrderQuantity],[SellerSupplierCustomerAssignedId],[SellerSupplierAdditionalId],[SellerSupplierPartyIdType],[SellerSupplierPartyId],[SellerSupplierPartyName],[LineValidityStartDate],[LineValidityStartTime],[LineValidityEndDate],[LineValidityEndTime],[LeadTimeUnit],[LeadTimeMeasure],[PriceMinimumQuantity],[PriceMaximumQuantity],[PriceAmountCurrency],[PriceAmount],[PriceBaseQuantityUnit],[PriceBaseQuantity],[PriceOrderableUnitFactorRate],[AllowanceChargeIndicator],[AllowanceChargeReasonCode],[AllowanceChargeReason],[AllowanceChargeMultiplierFactorNumeric],[AllowanceChargePrepaidIndicator],[AllowanceChargeAmount],[AllowanceChargeBaseAmount],[DeliveryUnitBatchQuantityUnit],[DeliveryUnitBatchQuantity],[DeliveryUnitConsumerQuantityUnit],[DeliveryUnitConsumerQuantity],[TaxCategoryId],[TaxCategoryName],[TaxCategoryPercent],[TaxCategoryBaseUnit],[TaxPerUnitAmount],[TaxExemptionReasonCode],[TaxExemptionReason],[TaxSchemeId],[TaxSchemeName],[TaxSchemeTypeListId],[TaxSchemeListAgencyId],[TaxSchemeListAgencyName],[TaxScheme],[TaxCurrencyCode],[ItemDescription],[PackQuantity],[PackSize],[ItemName],[HazardousRiskIndicator],[AdditionalInformation],[Keywords],[BrandName],[ModelName],[BuyersItemId],[SellersItemId],[ManufacturersItemId],[StandardItemId],[CatalogueItemId],[AdditionalItemId],[CommodityClassificationListName],[CommodityClassificationListVersion],[CommodityClassificationCode],[ManufacturerPartyIdType],[ManufacturerPartyId],[ManufacturerPartyName],[OriginAddressCountryId],[OriginAddressCountryName],[IsOsdItem],[PackLevel1Unit],[PackLevel1Quantity],[PackLevel1Ean],[PackLevel1Width],[PackLevel1Height],[PackLevel1Length],[PackLevel1GrossWeight],[PackLevel2Unit],[PackLevel2Quantity],[PackLevel2Ean],[PackLevel2Width],[PackLevel2Height],[PackLevel2Length],[PackLevel2GrossWeight],[PackLevel3Unit],[PackLevel3Quantity],[PackLevel3Ean],[PackLevel3Width],[PackLevel3Height],[PackLevel3Length],[PackLevel3GrossWeight],[PackLevel4Unit],[PackLevel4Quantity],[PackLevel4Ean],[PackLevel4Width],[PackLevel4Height],[PackLevel4Length],[PackLevel4GrossWeight],[PackLevelPalletUnit],[PackLevelPalletQuantity],[PackLevelPalletEan],[PackLevelPalletWidth],[PackLevelPalletHeight],[PackLevelPalletLength],[PackLevelPalletGrossWeight])
                                                      OUTPUT TempTable.ArrayIndex, ;
```

Please let me know if you need more information.

**Comment by andreisabau** [comment]
Hello @boginw 
It seems that your table doesn't have any identity column(s) which are generated by the database on insert.
Since the query is formed with those columns in mind (you can check it out here the code https://github.com/andreisabau/EFBulkInsert/blob/master/EFBulkInsert/BulkInsertExtension.cs#L100), the generatedColumnNames variable is empty and the query is ending in ",".
I should definitely add support for tables without any db generated columns.
In the meanwhile you could try to add/make a column as an identity, let me know if that works.

**Comment by boginw** [comment]
Hi @andreisabau 
I'm quite new to MSSQL and Entity Framework. Is it correct that identity column(s) are primary keys? If so, I have composite keys added like such `HasKey(ci => new { ci.CatalogueId, ci.LineId });`. Is this what you mean by Identity column(s)?

**Comment by andreisabau** [comment]
@boginw identity columns are different than primary keys.
You can read more here: https://dba.stackexchange.com/questions/46373/is-it-recommended-to-use-identity-as-an-alternative-to-primary-key or https://stackoverflow.com/questions/4293426/whats-the-difference-between-a-primary-key-and-identity - these explanation are better than what I could give.

What your table is missing (for EFBulkInsert to work in this version) is a column marked as an identity which basically auto increments the column's value on every INSERT.
I will also change EFBulkInsert to support tables without identity columns in a new version.

**Comment by andreisabau** [comment]
@boginw identity columns are different than primary keys.
You can read more here: https://dba.stackexchange.com/questions/46373/is-it-recommended-to-use-identity-as-an-alternative-to-primary-key or https://stackoverflow.com/questions/4293426/whats-the-difference-between-a-primary-key-and-identity - these explanations are better than what I could give.

What your table is missing (for EFBulkInsert to work in this version) is a column marked as an identity which basically auto increments the column's value on every INSERT.
From my knowledge the HasKey method makes the column as an identity only if it's used for a single key, not a composite one.
You can make other columns (int or bigint) as identity columns using the HasDatabaseGeneratedOption method or DatabaseGenerated attribute.
```
Property(x => x.ColumnName).HasDatabaseGeneratedOption(DatabaseGeneratedOption.Identity);

// OR

[DatabaseGenerated(DatabaseGeneratedOption.Identity)]
public int ColumnName { get; set;}
```

I will also change EFBulkInsert to support tables without identity columns in a new version.

**Comment by boginw** [comment]
I followed your instructions, and got a bit further. The bad news is that it still wont run. I've added an identity column like you suggested. I'm now getting an `System.Data.SqlClient.SqlException` exception with the message `Invalid object name 'CatalogueLineItem'`, which is understandable, since no table is called that. Instead the table is called `CatalogueLineItems`. Is there any way to specify the table name, or something similar?
